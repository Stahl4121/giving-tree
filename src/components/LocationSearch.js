import React from "react";

class Distance extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      centerLoc: {
        "lat" : 41.15559, 
        "long" : -80.08209
      },
      // lat and then long
       grantLocs: [
      {
        "lat": 36.8909, 
        "long" : -76.30892, 
        "name": "Bathroom Supplies", 
        "address": "1421 Bolling Ave, Norfolk, VA 23508"
      },
      {
        "lat": 38.89308, 
        "long": -76.97667, 
        "name": "Emotional Support", 
        "address": "1911 C Street NE Washington DC 20002"
      },
      {    
      "lat" : 41.16895,
      "long": -80.1132,
      "name": "Taco Bell",
      "address" : "1560 W. Main Street, Grove City, PA 16127"
      }
    ],
      dists: [],
    };
  }

  //from Geo Data Source
  calcDistance(lat1, lon1, lat2, lon2, unit) {
    if (lat1 === lat2 && lon1 === lon2) {
      return 0;
    } else {
      var radlat1 = (Math.PI * lat1) / 180;
      var radlat2 = (Math.PI * lat2) / 180;
      var theta = lon1 - lon2;
      var radtheta = (Math.PI * theta) / 180;
      var dist =
        Math.sin(radlat1) * Math.sin(radlat2) +
        Math.cos(radlat1) * Math.cos(radlat2) * Math.cos(radtheta);
      if (dist > 1) {
        dist = 1;
      }
      dist = Math.acos(dist);
      dist = (dist * 180) / Math.PI;
      dist = dist * 60 * 1.1515;
      if (unit === "K") {
        dist = dist * 1.609344;
      }
      if (unit === "N") {
        dist = dist * 0.8684;
      }
      return dist;
    }
  }

  fillDists() {  
    this.state.grantLocs.forEach(addDist); 
    function addDist(grantLoc) {
      this.setState(prevState => ({
        dists: [...prevState.dists,
                  {
                    "dist" : this.calcDistance(this.state.centerLoc.lat, this.state.centerLoc.long, this.state.grantLoc.lat, this.state.grantLoc.long, "M"),
                    "grantLoc" : grantLoc
                  }
                ]
      }))
      /*
      this.state.dists.push(
          {
              "dist" : this.calcDistance(this.state.centerLoc.lat, this.state.centerLoc.long, this.state.grantLoc.lat, this.state.grantLoc.long, "M"),
              "grantLoc" : grantLoc
          }
      );
      */
    }
    this.setState(prevState => ({
      dists: prevState.dists.sort((a, b) => (a.dist > b.dist ? 1 : -1))
    }))
    /*
    this.setState({
      dists: this.state.dists.sort((a, b) => (a.dist > b.dist ? 1 : -1))
    }) 
    */
  }
}
/*

}
//from Geo Data Source
function calcDistance(lat1, lon1, lat2, lon2, unit) {
  if (lat1 === lat2 && lon1 === lon2) {
    return 0;
  } else {
    var radlat1 = (Math.PI * lat1) / 180;
    var radlat2 = (Math.PI * lat2) / 180;
    var theta = lon1 - lon2;
    var radtheta = (Math.PI * theta) / 180;
    var dist =
      Math.sin(radlat1) * Math.sin(radlat2) +
      Math.cos(radlat1) * Math.cos(radlat2) * Math.cos(radtheta);
    if (dist > 1) {
      dist = 1;
    }
    dist = Math.acos(dist);
    dist = (dist * 180) / Math.PI;
    dist = dist * 60 * 1.1515;
    if (unit === "K") {
      dist = dist * 1.609344;
    }
    if (unit === "N") {
      dist = dist * 0.8684;
    }
    return dist;
  }
}

// props includes
// 1. a location (centerLoc)
// 2. an array of the grant locations (grantLocs)
export default function getDistance(props) {
  // lat and then long
  const grantLocs = [
    
    {
        "lat": 36.8909, 
        "long" : -76.30892, 
        "name": "Bathroom Supplies", 
        "address": "1421 Bolling Ave, Norfolk, VA 23508"
    },
    {
        "lat": 38.89308, 
        "long": -76.97667, 
        "name": "Emotional Support", 
        "address": "1911 C Street NE Washington DC 20002"
    },
    {    
      "lat" : 41.16895,
      "long": -80.1132,
      "name": "Taco Bell",
      "address" : "1560 W. Main Street, Grove City, PA 16127"
    }
  ];

  //var myHouse = [36.8909, -76.30892];
  var dists = [];
  var centerLoc = {
      "lat" : 41.15559, 
      "long" : -80.08209
  };
  
  grantLocs.forEach(addDist); 
  function addDist(grantLoc) {
    dists.push(
        {
            "dist" : calcDistance(centerLoc.lat, centerLoc.long, grantLoc.lat, grantLoc.long, "M"),
            "grantLoc" : grantLoc
        }
    );
  } 
  dists.sort((a, b) => (a.dist > b.dist ? 1 : -1));
  return (
    <div>
        <div>
            {JSON.stringify(dists)}
        </div>
    </div>);
}

get the central location
store it

get the
*/
