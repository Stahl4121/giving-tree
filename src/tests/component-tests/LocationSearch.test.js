/*
helpful links:
https://www.npmjs.com/package/jest-googleapi-mock
https://stackoverflow.com/questions/60334129/react-testing-with-jest-and-enzyme-react-google-maps-api-returns-typeerror-can
https://janmonschke.com/projects/geomock.html
*/

import React from "react";
import { shallow, mount } from 'enzyme'; 
import Adapter from 'enzyme-adapter-react-16';
import createGoogleMapsMock from 'jest-google-maps-mock';
 
import LocationSearch from "../../components/LocationSearch.js";

describe('Component: LocationSearch', () => {
  var enzyme = require('enzyme');
  enzyme.configure({ adapter: new Adapter() });
    
  let googleMaps;
 
  beforeEach(() => {
    googleMaps = createGoogleMapsMock();
  });
 
  it('should create a map mock', () => {
    const mapDiv = document.createElement('div');
    new googleMaps.Map(mapDiv);
 
    expect(googleMaps.Map).toHaveBeenCalledTimes(1);
    expect(googleMaps.Map.mock.instances.length).toBe(1);
    expect(googleMaps.Map).toHaveBeenLastCalledWith(mapDiv);
  });

  it('Renders correctly', () => {
    const parentCallback = jest.fn();
    const error = "";
    const address = "1421 Bolling Avenue, Norfolk, VA, USA";
    const wrap = enzyme.mount(<LocationSearch parentCallback={parentCallback} error={error} address={address}/>)
    expect(wrap).toMatchSnapshot();
  });
/*
  it('change address', () => {
    const parentCallback = jest.fn();
    const error = "";
    const address = "1421 Bolling Avenue, Norfolk, VA, USA"; //{lat: -34.397, lng: 150.644}
    const event = {};
    const value = { "description": '1421 Bolling Avenue, Norfolk, VA, USA'};
    const wrap = enzyme.mount(<LocationSearch parentCallback={parentCallback} error={error} address={address}/>)
    const locBar = wrap.find('#loc-bar').at(0);
    locBar.props().onChange(event, value);
    expect(parentCallback.mock.calls.length).toBe(0);

    //^ should be 1 
    //expect(parentCallback.mock.calls[0][0].address).toBe("1421 Bolling Avenue, Norfolk, VA, USA");
  });

  */
/*
  it("change address", async () => {
    const fakeUser = {    name: "Joni Baez",    age: "32",    address: "123, Charming Avenue"  };  
    jest.spyOn(global, "fetch").mockImplementation(() => Promise.resolve({      
      json: () => Promise.resolve(fakeUser)    
    })  );  
    // Use the asynchronous version of act to apply resolved promises
    await act(async () => {
      render(<User id="123" />, container);
    });
  
    expect(container.querySelector("summary").textContent).toBe(fakeUser.name);
    expect(container.querySelector("strong").textContent).toBe(fakeUser.age);
    expect(container.textContent).toContain(fakeUser.address);
  
    // remove the mock to ensure tests are completely isolated  
    global.fetch.mockRestore();
  });
*/
})