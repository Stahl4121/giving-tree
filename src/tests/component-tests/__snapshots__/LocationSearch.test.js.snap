// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: LocationSearch Renders correctly 1`] = `ReactWrapper {}`;
